Added the base template

ElenaKusevska opened the pull request:
@rgaudin please review this PR, and let me know if I should make any changes.

ElenaKusevska committed:
add base.html and css.html

ElenaKusevska committed:
Add Jinja2 templating. Edit setup.py to copy templates and assets

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -31,7 +33,20 @@ def __init__(self, **kwargs):
+
+        template = self.env.get_template("base.html")
+
+        logger.info(template.render(title="base", description="d"))
```
We want a Zim file with a home page at this stage.

rgaudin commented on wikihow2zim/templates/base.html:
```diff
@@ -0,0 +1,44 @@
+        <title>
+            {{title}}
+        </title>
+        <link href="https://www.wikihow.com/skins/WikiHow/wH-initials_152x152.png" rel="apple-touch-icon"/>
```
No link to online resource. Add to zim.

rgaudin commented on wikihow2zim/templates/base.html:
```diff
@@ -0,0 +1,44 @@
+        <!-- Header -->
+        <div id="header_container">
+            <div id="header" >
+                <a href="{{ to_root }}templates/base.html">
```
templates should not appear in the path

rgaudin commented on wikihow2zim/templates/base.html:
```diff
@@ -0,0 +1,44 @@
+                {% block footer_breadcrumbs %}
+                {% endblock %}
+                <div id="footer_help">
+                    <a href="{{ to_root }}templates/categories.html"> List Categories </a>
```
Idem for templates in path

rgaudin reviewed (changes requested):
Thank you @ElenaKusevska  but I'm not sure what this is trying to achieve.

Issue #3 should create a Zim file with a somewhat blank home page that uses a base template, itself including a header and footer.

- fix the codefactor issues if any (or mention it might be a false positive)
- Logo should be donwloaded by the scraper and added to Zim
- There should be no reference to online resource (`<link href="https://www.wikihow.com/skins/WikiHow/wH-initials_152x152.png"`)
- the base.css file seems to be included twice
- templates folder should not appear in in-zim links (`<a href="{{ to_root }}templates/base.html">`)

And please, don't push commits to a PR after you've requested a review, or un-request the review or mark it as draft. It's annoying to have to re-check what's change while in review.

ElenaKusevska commented:
@rgaudin, Oh sorry, I didn't realize you wanted to create the zim for this issue. OK. I'll do that now. Sorry about pushing, I'll keep that in mind.

rgaudin commented:
Thanks 😉

ElenaKusevska committed:
create "/Home" zim entry using base template

ElenaKusevska commented:
@rgaudin, have a look at these changes. Let me know if this is more similar to what you had in mind, and what additional changes you would recommend.

rgaudin commented on wikihow2zim/entrypoint.py:
```diff
@@ -47,7 +47,7 @@ def main():
         "--output",
         help="Output folder for ZIM file",
-        default="/output",
+        default="output",
```
We use `/output` as default on all scrapers because we distribute mainly through docker and it's easier to have a predictable path to advertise mounting point. I reckon it's a bit annoying when using locally but let's keep it as `/output` until we decide of a better default and apply that to all scrapers.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -1,6 +1,15 @@
 
-from .constants import URLS, getLogger
+import os
+import bs4
```
Separate imports in three groups: stdlib, 3rd-parties, internal

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
         for option in options:
             setattr(self, option, kwargs[option])
 
+        if self.output_dir is None:
```
This should never be `None` as you have a default coming from argparse.

ElenaKusevska replied:
@rgaudin You're right :) I don't know how I didn't see that.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
 
+        # jinja2 environment setup
+        self.env = Environment(
+            loader=FileSystemLoader(str(self.templates_dir)), autoescape=True
```
Why the `str()`. Jinja understands pathlib.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+        )
+
+        # Set up the output zim file:
+        fpath = pathlib.Path(self.output_dir).joinpath("output.zim")
```
Not much value in creating this variable just to use it once on next line.
You may want to keep that as an instance prop though if you want to display it later on in the process.

ElenaKusevska replied:
@rgaudin The only reason I have it like that is that the code looks more readable to me. Otherwise the line is too long and if I try to have it in several lines it doesn't look good. Creating a single variable like this isn't particularly computationally expensive, but we can change it to be in one line if you think it's better. Either way is fine with me.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+
+        # Set up the output zim file:
+        fpath = pathlib.Path(self.output_dir).joinpath("output.zim")
+        self.creator = Creator(filename=fpath).set_mainpath("Home")
```
We never create nor start the Zim Creator in `__init__()` because libzim's a C++ extension and that comes with a few constraints ; mostly in error reporting.
For that reason, you'll want to create and start it in your `run()` method, after you're done checking inputs.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+        return ROOT_DIR.joinpath("templates")
+
+    def add_assets(self):
+        assets_path = ROOT_DIR.joinpath("assets")
```
No need to create a var here just to use it once

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+
+    def add_assets(self):
+        assets_path = ROOT_DIR.joinpath("assets")
+        for root, dirs, files in os.walk(assets_path):
```
`dirs` is not used. Don't declare it.
We also prefer using pathlib's helpers rather than `os` ones, when possible.
You may also want to add a debug print

```py
assets_root = ROOT_DIR.joinpath("assets")
for fpath in assets_root.glob("**/*"):
    if not fpath.is_file():
        continue
    logger.debug(f"Assed {fpath.relative_to(assets_root)}")
    self.creator.add_item_for(path=str(fpath.relative_to(assets_root)), fpath=fpath)
````

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+        )
+
+        # Add wikihow logo:
+        LOGO_PATH = "https://www.wikihow.com/extensions/wikihow/mobile/images/wikihow_logo_230.png"
```
It's not a path, it's an url. Also, don't use all-caps as this would be mistaken for a global variable.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+
+        # Add wikihow logo:
+        LOGO_PATH = "https://www.wikihow.com/extensions/wikihow/mobile/images/wikihow_logo_230.png"
+        logo_path = ROOT_DIR.joinpath("assets", "static", "wikihow_logo.png")
```
See, you have two variables with same name, but different cases. That's confusing.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+        LOGO_PATH = "https://www.wikihow.com/extensions/wikihow/mobile/images/wikihow_logo_230.png"
+        logo_path = ROOT_DIR.joinpath("assets", "static", "wikihow_logo.png")
+        self.creator.add_item(
+            URLItem(url=LOGO_PATH, path="logo_path", mimetype="image/png")
```
You're adding that image at `/logo_path` in the Zim, not at `/assets/static/wikihow_logo.png` because that's a string and not a variable you're using

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
+            URLItem(url=LOGO_PATH, path="logo_path", mimetype="image/png")
+        )
+
+        # fetch icon from website and set it as Zim Icon
```
All of the above is related to the illustration, not the homepage. That should be moved.

rgaudin commented on wikihow2zim/scraper.py:
```diff
@@ -29,9 +38,73 @@ def __init__(self, **kwargs):
     def run(self):
         logger.info(
-            f"running the scraper with the following arguments: \n{self.__dict__}"
+            f"\nrunning the scraper with the following arguments: \n\n{self.__dict__}"
```
Why the extra `\n` ? It disconnects the message from the log prefix. Also we can get rid of the `.__dict__` now.

rgaudin reviewed (changes requested):
Thank you @ElenaKusevska. I've added some inline change requests ; please have a look. Also:

- Remove the wikihow logo from the repo.
- I am attaching a mockup of what the base template could look like. Yours require scrolling vertically while there is no content and the footer is not that appealing.

Additional tip: use the PR description to briefly explains what you've done so reviewer knows what to expect. Don't do it for this PR, I've already reviewed and the task it's implementing make it quite obvious. Just a general comment.

**Yours**
![Home](https://user-images.githubusercontent.com/57929/129009002-af660452-0dca-403d-88c2-0e9ca36fecd4.png)
**Mockup**
![Screen Shot 2021-08-11 at 09 25 30](https://user-images.githubusercontent.com/57929/129009025-7011883b-b2e4-4c96-8c6d-69b399aa9246.png)

ElenaKusevska commented:
@rgaudin check if it's OK now.

Commit:
Fixed formatting

- No trailing white spaces
- No unused imports
- lines wrapped at 88 chars (black)
- No f-string without actual need

Python-related ones are provided by flake8

Commit:
Moved icon code out of run()

`run()` is the main entry point of the scraper and will launch all the sequential steps.
It shouldn't be polluted with such code which decreases readingness and increase
complexity. Also, we don't need its variables to be kept alive all this time

rgaudin reviewed (approved):
Good. I fixed a couple stuff directly. Please take a look at those commits as we'd want that going forward.
